cluster-autoscaler/core/static_autoscaler_dra_test.go

```diff
@@ -0,0 +1,777 @@
+			// The claim is bound to a node, and the node only fits 10 pods because of CPU. The 3 extra pods shouldn't trigger a scale-up.
+			expectedNoScaleUps: []noScaleUpDef{{podNamePrefix: "unschedulable", podCount: 3}},
+		},
+		"scale-down: empty single-device nodes": {
+			nodeGroups: map[*testNodeGroupDef]int{node1GpuA1slice: 3},
+			pods: []testPod{
+				scheduledPod(baseBigPod, "scheduled-0", node1GpuA1slice.name+"-1", map[*testDeviceRequest][]string{&req1GpuA: {gpuDevice + "-0"}}),
+			},
+			expectedScaleDowns:   map[string][]string{node1GpuA1slice.name: {node1GpuA1slice.name + "-0", node1GpuA1slice.name + "-2"}},
+			expectedNoScaleDowns: []noScaleDownDef{{nodeName: node1GpuA1slice.name + "-1", reason: simulator.NotUnderutilized}},
+		},
+		"scale-down: single-device nodes with drain": {
+			nodeGroups: map[*testNodeGroupDef]int{node3GpuA1slice: 3},
+			pods: []testPod{
+				scheduledPod(baseSmallPod, "scheduled-0", node3GpuA1slice.name+"-0", map[*testDeviceRequest][]string{&req2GpuA: {gpuDevice + "-0", gpuDevice + "-1"}}),
+				scheduledPod(baseSmallPod, "scheduled-1", node3GpuA1slice.name+"-1", map[*testDeviceRequest][]string{&req2GpuA: {gpuDevice + "-0", gpuDevice + "-1"}}),
+				scheduledPod(baseSmallPod, "scheduled-2", node3GpuA1slice.name+"-2", map[*testDeviceRequest][]string{&req1GpuA: {gpuDevice + "-0"}}),
+				scheduledPod(baseSmallPod, "scheduled-3", node3GpuA1slice.name+"-2", map[*testDeviceRequest][]string{&req1GpuA: {gpuDevice + "-1"}}),
+			},
+			expectedScaleDowns: map[string][]string{node3GpuA1slice.name: {node3GpuA1slice.name + "-2"}},
+			expectedNoScaleDowns: []noScaleDownDef{
+				{nodeName: node1GpuA1slice.name + "-0", reason: simulator.NoPlaceToMovePods},
+				{nodeName: node1GpuA1slice.name + "-1", reason: simulator.NoPlaceToMovePods},
+			},
+		},
+		"no scale-down: no place to reschedule": {
+			nodeGroups: map[*testNodeGroupDef]int{node3GpuA1slice: 3},
+			pods: []testPod{
+				scheduledPod(baseBigPod, "scheduled-0", node3GpuA1slice.name+"-0", map[*testDeviceRequest][]string{&req2GpuA: {gpuDevice + "-0", gpuDevice + "-1"}}),
+				scheduledPod(baseBigPod, "scheduled-1", node3GpuA1slice.name+"-1", map[*testDeviceRequest][]string{&req2GpuA: {gpuDevice + "-0", gpuDevice + "-1"}}),
+				scheduledPod(baseBigPod, "scheduled-2", node3GpuA1slice.name+"-1", map[*testDeviceRequest][]string{&req1GpuA: {gpuDevice + "-2"}}),
+				scheduledPod(baseSmallPod, "scheduled-3", node3GpuA1slice.name+"-2", map[*testDeviceRequest][]string{&req1GpuA: {gpuDevice + "-0"}}),
+				scheduledPod(baseSmallPod, "scheduled-4", node3GpuA1slice.name+"-2", map[*testDeviceRequest][]string{&req1GpuA: {gpuDevice + "-1"}}),
+			},
+			expectedNoScaleDowns: []noScaleDownDef{
+				{nodeName: node3GpuA1slice.name + "-0", reason: simulator.NoPlaceToMovePods},
+				{nodeName: node3GpuA1slice.name + "-1", reason: simulator.NotUnderutilized},
+				{nodeName: node3GpuA1slice.name + "-2", reason: simulator.NoPlaceToMovePods},
+			},
+		},
```

Comment:
https://github.com/kubernetes/autoscaler/pull/7530/#issuecomment-2536807564
> Integration tests (7th commit) - do the scenarios listed there make sense and behave as you would expect them to? Are some major scenarios missing?

A scale down scenario where a pod using a claim gets moved successfully to some other node would be good. As discussed in https://github.com/kubernetes/autoscaler/pull/7530/#discussion_r1882417954, this probably doesn't work yet unless deallocation of the claim gets added. That scenario then also can cover utilization calculation (the "least utilized" node is meant to get removed).

A [monitoring DS with admin mode](https://github.com/kubernetes/autoscaler/pull/7530/#discussion_r1882368788) might also be useful, once support for that gets added.

Specific scenarios aside, a general comment about testing: I see that DRA support is behind a flag. Would it be possible to run all existing test scenarios with and without that flag? Nothing should change, but one cannot be sure unless it gets tested...

The inverse is also true: what happens in a cluster where DRA is used and the flag is off? The outcome is probably different from the "flag is on" case, but is it still sane?

Reply:
> A scale down scenario where a pod using a claim gets moved successfully to some other node would be good. As discussed in #7530, this probably doesn't work yet unless deallocation of the claim gets added. That scenario then also can cover utilization calculation (the "least utilized" node is meant to get removed).

Doesn't "scale-down: single-device nodes with drain" cover that? Deallocating claims is already implemented, I was just wondering if it was the correct behavior to simulate.

> A monitoring DS with admin mode might also be useful, once support for that gets added.

Definitely, these tests should be expanded for the post-MVP features.

> Specific scenarios aside, a general comment about testing: I see that DRA support is behind a flag. Would it be possible to run all existing test scenarios with and without that flag? Nothing should change, but one cannot be sure unless it gets tested...
> 
> The inverse is also true: what happens in a cluster where DRA is used and the flag is off? The outcome is probably different from the "flag is on" case, but is it still sane?

* So all existing pre-DRA testing scenarios are run with the DRA flag enabled, to cover more code paths and ensure that previous functionality still works as expected with DRA enabled but no DRA objects to work on.
* I agree that ideally we'd also run all of them with DRA disabled and no DRA objects to work on for 100% certainty that the behavior stays the same, but that seems hard to do technically. `PredicateSnapshot()` takes the DRA flag, and it's used in unit tests throughout the codebase. Right now all these tests just call `testsnapshot.TestClusterSnapshotOrDie(t)` to get a test snapshot (with DRA enabled as described above), and I think they would all need to be modified to iterate over both flag options. This seems like a lot of work for slightly more test coverage.
* Now the "DRA is disabled but there are DRA objects" is something that's indeed not well covered. I don't think covering it is critical for MVP, added a TODO for post-MVP. WDYT?

Reply:
> Doesn't "scale-down: single-device nodes with drain" cover that? Deallocating claims is already implemented, I was just wondering if it was the correct behavior to simulate.

Then we are good.

> This seems like a lot of work for slightly more test coverage.

Ack.

> Now the "DRA is disabled but there are DRA objects" is something that's indeed not well covered. I don't think covering it is critical for MVP, added a TODO for post-MVP. WDYT?

I'm okay with extending code coverage later.